### PR TITLE
Add systemd notify and watchdog support

### PR DESCRIPTION
**Pull request by ekohl**
### Description
This is a rebased version of https://github.com/puma/puma/pull/2260 with additional changes. I squashed all commits from the other PR into a single one to keep credits for @acmh while also making it easy to rebase. All my changes are on top of it.

It is more than 169 lines change, but splitting it up might be hard with credit. If desired, I can first submit the additional events as a separate patch.

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by nateberkopec** [comment]
> It is more than 169 lines change, but splitting it up might be hard with credit. 

That's fine. I appreciate the effort and desire to share the credit :)

**Comment by ekohl** [comment]
Added a `skip_on :jruby` to the tests since that doesn't work. At least a WIP test passed so let's see the full set now.

**Comment by nateberkopec** [comment]
Hmmm... any idea why it doesn't work with JRuby? It _should_ in theory, right?

**Comment by ekohl** [comment]
It gave an error on bind with the unix socket, I'm not sure why. https://github.com/puma/puma/blob/master/docs/systemd.md#socket-activation does state:
> Note: Socket activation doesn't currently work on JRuby. This is tracked in #1367.

Following the trail of issues it looks like this is still an issue on JRuby.

File: lib/puma/systemd.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+          sleep ping_f
+          SdNotify.watchdog
+        end
+      end
+    end
+
+    private
+
+    def watchdog_sleep_time
+      return unless SdNotify.watchdog?
```
Comment (ehelms) [inline lib/puma/systemd.rb:33]: I think it would improve the understanding here to move this up to just after the `def start_watchdog` line.

Comment (ekohl) [reply lib/puma/systemd.rb]: Agreed, moved.

**Comment by nateberkopec** [comment]
LGTM but would appreciate review from others. We'll merge this in after 5.0.3 releases.

**Comment by acmh** [comment]
Congratz @ekohl , thank you for the credit! Good work!

**Comment by ekohl** [comment]
Thanks for the initial work @acmh! It was a very solid base to improve on.

**Comment by rromanchuk** [comment]
I think i have a weird configuration issue using type=notify, everything starts up fine, serving request and then systemd kills at the WatchdogSec interval. I'll come back with better details after i poke around some more. Maybe i missed something with sd_notify 

```
puma.service: start operation timed out. Terminating.
puma[19721]: [19721] - Gracefully shutting down workers...
puma[20181]: On worker shutdown...
puma[20182]: On worker shutdown...
systemd[1]: puma.service: Failed with result 'timeout'.
```

File: lib/puma/systemd.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      @events.on_restart { SdNotify.reloading }
+    end
+
+    def start_watchdog
+      return unless SdNotify.watchdog?
+
+      ping_f = watchdog_sleep_time
+
+      log "Pinging systemd watchdog every #{ping_f.round(1)} sec"
+      Thread.new do
```
Comment (schuetzm) [inline lib/puma/systemd.rb:23]: Hmm... doesn't it defeat the purpose if you start the watchdog as a separate thread? How would that ever fail? It should either be integrated into the main event loop or at least somehow check whether the service is actually still working.

Comment (ekohl) [reply lib/puma/systemd.rb:23]: I think that's a fair point but I wouldn't know how to do it better. Perhaps this is a good enough implementation for now and it allows for a better implementation in the future. Maybe the thread should perform a request to see if it's served? The question then becomes: which one? Should it be configurable?

**Comment by ekohl** [comment]
> I think i have a weird configuration issue using type=notify, everything starts up fine, serving request and then systemd kills at the WatchdogSec interval. I'll come back with better details after i poke around some more. Maybe i missed something with sd_notify

Did you figure it out? Perhaps you should start without the watchdog to see if that does work.

**Comment by Dantemss** [comment]
@rromanchuk I just ran into a similar issue with a different daemon... if your `ExecStart` uses a shell, make sure you `exec` the puma command rather than just calling it so that systemd gets the correct process as the main process (otherwise you have both a shell and puma processes, systemd thinks the shell is the main process and for whatever reason that makes `SdNotify.watchdog?` return false)

**Comment by mperham** [comment]
With Sidekiq I decided to vendor the code for the `sd_notify` gem because it's only ~50 lines of stable code. I didn't want to require the user to add the gem to their gemfile manually, that shouldn't be something they need to care about IMO.

**Pull request merged**